src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,508 @@
+
+    // Set new data. All fresh new qubits are added to the free qubits in the outermost area.
+    freeQubitCount += requestedCapacity - qubitCapacity;
+    FailIf(freeQubitCount <= 0, "Incorrect free qubit count.");
+    delete[] sharedQubitStatusArray;
+    sharedQubitStatusArray = newStatusArray;
+    qubitCapacity = requestedCapacity;
+    freeQubitsInAreas[0].FreeQubitsReuseAllowed.MoveAllQubitsFrom(newFreeQubits, sharedQubitStatusArray);
+}
+
+CQubitManager::QubitIdType CQubitManager::AllocateQubitId()
+{
+    QubitIdType id = NoneMarker;
+    if (encourageReuse)
+    {
+        // When reuse is encouraged, we start with the innermost area
+        for (CRestrictedReuseAreaStack::reverse_iterator rit = freeQubitsInAreas.rbegin(); rit != freeQubitsInAreas.rend(); ++rit)
+        {
+            id = rit->FreeQubitsReuseAllowed.TakeQubitFromFront(sharedQubitStatusArray);
+            if (id != NoneMarker)
+            {
+                break;
+            }
+        }
+    } else
+    {
+        // When reuse is discouraged, we start with the outermost area
+        for (CRestrictedReuseAreaStack::iterator it = freeQubitsInAreas.begin(); it != freeQubitsInAreas.end(); ++it)
+        {
+            id = it->FreeQubitsReuseAllowed.TakeQubitFromFront(sharedQubitStatusArray);
+            if (id != NoneMarker)
+            {
+                break;
+            }
+        }
+    }
+    if (id != NoneMarker) {
+        FailIf(id < 0 || id >= qubitCapacity, "Internal Error: Allocated invalid qubit.");
+        allocatedQubitCount++;
+        FailIf(allocatedQubitCount <= 0, "Incorrect allocated qubit count.");
+        freeQubitCount--;
+        FailIf(freeQubitCount < 0, "Incorrect free qubit count.");
+    }
+    return id;
+}
+
+void CQubitManager::ReleaseQubitId(QubitIdType id)
+{
+    FailIf(id < 0 || id >= qubitCapacity, "Internal Error: Cannot release an invalid qubit.");
+    if (IsDisabled(id))
+    {
+        // Nothing to do. Qubit will stay disabled.
+        return;
+    }
+
+    FailIf(!IsExplicitlyAllocated(id), "Attempt to free qubit that has not been allocated.");
+
+    if (mayExtendCapacity && !encourageReuse)
+    {
+        // We can extend capcity and don't want reuse => Qubits will never be reused => Discard qubit.
+        // We put it in its own "free" list, this list will never be found again and qubit will not be reused.
+        sharedQubitStatusArray[id] = NoneMarker;
+    } else
+    {
+        // Released qubits are added to reuse area/segment in which they were released
+        // (rather than area/segment where they are allocated).
+        // Although counterintuitive, this makes code simple.
+        // This is reasonable because qubits should be allocated and released in the same segment. (This is not enforced)
+        freeQubitsInAreas.PeekBack().FreeQubitsReuseAllowed.AddQubit(id, encourageReuse, sharedQubitStatusArray);
+    }
+
+    freeQubitCount++;
```

Comment:
I have a suspicion that `freeQubitCount++` (line 486) should only happen in the `else` part (lines 478 - 484), i.e. only if the qubit will be _reused_.

Reply:
This is a great observation and a great question. It comes down to the definition of what freeQubitCount is, which is, unfortunately, not provided anywhere! freeQubitCount is the count of free qubits tracked by qubit manager. Technically these qubits are tracked so increase is always needed. Similar question may be asked - why freeQubitCount is let's say 4, when qubit manager may extend capacity? I will add comments.